sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelUtils.java

```diff
@@ -51,18 +62,117 @@
    *
    * <p>For example, when reading from "gs://bucket/path", the scheme "gs" is
    * used to lookup the appropriate factory.
+   *
+   * @throws RuntimeException if override is false and schemes are conflicting.
    */
-  public static void setIOFactory(String scheme, IOChannelFactory factory) {
+  @VisibleForTesting
+  public static void setIOFactoryInternal(
+      String scheme,
+      IOChannelFactory factory,
+      boolean override) {
+    if (!override && FACTORY_MAP.containsKey(scheme)) {
+      throw new RuntimeException(String.format(
+          "Failed to register IOChannelFactory: %s. "
+              + "Scheme: [%s] is already registered with %s, and override is not allowed.",
+          FACTORY_MAP.get(scheme).getClass(),
+          scheme,
+          factory.getClass()));
+    }
     FACTORY_MAP.put(scheme, factory);
   }
 
   /**
+   * Deregisters the scheme and the associated {@link IOChannelFactory}.
+   */
+  @VisibleForTesting
+  static void deregisterScheme(String scheme) {
+    FACTORY_MAP.remove(scheme);
+  }
+
+  /**
    * Registers standard factories globally. This requires {@link PipelineOptions}
    * to provide, e.g., credentials for GCS.
+   *
+   * @deprecated use {@link #registerIOFactories}.
    */
+  @Deprecated
   public static void registerStandardIOFactories(PipelineOptions options) {
-    setIOFactory("gs", new GcsIOChannelFactory(options.as(GcsOptions.class)));
-    setIOFactory("file", new FileIOChannelFactory());
+    registerIOFactoriesAllowOverride(options);
+  }
+
+  /**
+   * Registers all {@link IOChannelFactory IOChannelFactories} from {@link ServiceLoader}.
+   *
+   * <p>This requires {@link PipelineOptions} to provide, e.g., credentials for GCS.
+   *
+   * <p>Override existing schemes is not allowed.
+   *
+   * @throws RuntimeException if the scheme of a {@link IOChannelFactory} is already registered.
```

Comment:
Throw IllegalStateException since it better correlates with the issue.

@throws IllegalStateException if multiple {@link IOChannelFactory}s for the same scheme are detected.